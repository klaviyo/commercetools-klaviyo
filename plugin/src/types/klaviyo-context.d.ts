import { KlaviyoService } from '../infrastructure/driven/klaviyo/KlaviyoService';
import { OrderMapper } from '../domain/shared/mappers/OrderMapper';
import { CustomerMapper } from '../domain/shared/mappers/CustomerMapper';
import { CategoryMapper } from '../domain/shared/mappers/CategoryMapper';
import { ProductMapper } from '../domain/shared/mappers/ProductMapper';
import { CtCustomerService } from '../infrastructure/driven/commercetools/CtCustomerService';
import { CtProductService } from '../infrastructure/driven/commercetools/CtProductService';
import { CtCategoryService } from '../infrastructure/driven/commercetools/CtCategoryService';
import { CtPaymentService } from '../infrastructure/driven/commercetools/CtPaymentService';
import { CtOrderService } from '../infrastructure/driven/commercetools/CtOrderService';

type Context = {
    klaviyoService: KlaviyoService;
    orderMapper: OrderMapper;
    customerMapper: CustomerMapper;
    categoryMapper: CategoryMapper;
    productMapper: ProductMapper;
    ctCustomerService: CtCustomerService;
    ctProductService: CtProductService;
    ctCategoryService: CtCategoryService;
    ctPaymentService: CtPaymentService;
    ctOrderService: CtOrderService;
};
