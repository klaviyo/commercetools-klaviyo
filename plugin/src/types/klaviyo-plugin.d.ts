type KlaviyoEvent = {
    type:
        | 'profileCreated'
        | 'profileUpdated'
        | 'profileResourceUpdated'
        | 'categoryCreated'
        | 'categoryDeleted'
        | 'categoryUpdated'
        | 'itemCreated'
        | 'itemUpdated'
        | 'itemDeleted'
        | 'variantCreated'
        | 'variantUpdated'
        | 'variantDeleted'
        | 'event';
    body: KlaviyoRequestType;
};

type KlaviyoRequestType =
    | ProfileRequest
    | EventRequest
    | CategoryRequest
    | CategoryDeletedRequest
    | ItemRequest
    | ItemVariantRequest
    | ItemJobRequest
    | ItemVariantJobRequest
    | ItemDeletedRequest;

type ProcessingResult = {
    status: 'OK' | '4xx';
};

type maxSizeJobArrayContainer = {
    [key: string]: KlaviyoEvent[];
};
