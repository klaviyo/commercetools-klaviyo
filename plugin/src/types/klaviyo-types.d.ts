type ProfileRequest = {
    data: ProfileType;
};

type ProfileType = {
    type: 'profile';
    id?: string;
    attributes: Profile;
    links?: any;
    relationships?: any;
    meta?: {
        identifiers: {
            external_id: string;
        };
    };
};

type Profile = {
    external_id?: string;
    anonymous_id?: string;
    email?: string;
    first_name?: string;
    last_name?: string;
    title?: string;
    phone_number?: string | null;
    organization?: string;
    image?: string | null;
    created?: string;
    updated?: string;
    last_event_date?: string;
    location?: KlaviyoLocation | null;
    properties?: any;
};

type KlaviyoLocation = {
    address1?: string | null;
    address2?: string | null;
    city?: string;
    country?: string;
    latitude?: string | null;
    longitude?: string | null;
    region?: string;
    zip?: string;
    timezone?: string | null;
};

type EventRequest = {
    data: EventType;
};

type EventType = {
    type: 'event';
    id?: string;
    attributes: {
        profile: KlaviyoEventProfile;
        metric: {
            name: string;
        };
        value: number;
        properties: any;
        unique_id: string;
        time: string;
    };
};

type KlaviyoEventProfile = {
    $email?: string;
    $id?: string;
    $first_name?: string;
    $last_name?: string;
    $phone_number?: string;
    [key: string]: unknown;
};

type CategoryRequest = {
    data: CategoryType;
};

type CategoryType = {
    type: 'catalog-category';
    id?: string;
    attributes: KlaviyoCategory;
    relationships?: KlaviyoRelationships;
};

type KlaviyoCategory = {
    external_id?: string;
    id?: string;
    name: string;
    integration_type?: string | '$custom';
    catalog_type?: string | '$default';
    updated?: string;
};

type CategoryDeletedRequest = {
    data: {
        id: string;
    };
};

type ItemRequest = {
    data: ItemType;
    variantJobRequests?: KlaviyoEvent[];
};

type ItemType = {
    type: 'catalog-item';
    id?: string;
    attributes: KlaviyoCatalogItem;
    relationships?: KlaviyoRelationships;
};

type KlaviyoCatalogItem = {
    external_id?: string;
    id?: string;
    integration_type?: string | '$custom';
    catalog_type?: string | '$default';
    title?: string;
    description?: string;
    url?: string;
    image_full_url?: string;
    published: boolean;
    price?: number;
    custom_metadata?: any;
};

type ItemVariantRequest = {
    data: ItemVariantType;
};

type ItemVariantType = {
    type: 'catalog-variant';
    id?: string;
    attributes: KlaviyoCatalogVariant;
    relationships?: KlaviyoRelationships;
};

type ItemDeletedRequest = {
    data: {
        id: string;
        deleteVariantsJob?: KlaviyoEvent;
    };
};

type KlaviyoCatalogVariant = {
    external_id?: string;
    id?: string;
    integration_type?: string | '$custom';
    catalog_type?: string | '$default';
    title?: string;
    description?: string;
    sku?: string;
    inventory_quantity?: number;
    inventory_policy?: 0 | 1 | 2;
    price?: number;
    url?: string;
    image_full_url?: string;
    published: boolean;
    custom_metadata?: any;
};

type KlaviyoRelationships = {
    [key: string]: {
        data: KlaviyoRelationshipData[]
    }
}

type KlaviyoRelationshipData = {
    type: 'catalog-item' | 'catalog-variant' | 'catalog-category';
    id: string;
}

type ItemJobRequest = {
    data: ItemJobType;
};

type ItemJobType = {
    type: 'catalog-item-bulk-create-job' | 'catalog-item-bulk-update-job';
    id?: string,
    attributes: {
        items: ItemType[];
    };
};

type ItemVariantJobRequest = {
    data: ItemVariantJobType;
};

type ItemVariantJobType = {
    type: 'catalog-variant-bulk-create-job' | 'catalog-variant-bulk-update-job' | 'catalog-variant-bulk-delete-job';
    id?: string,
    attributes: {
        variants: ItemVariantType[];
    };
};

type KlaviyoQueryResult<KlaviyoElement> = {
    data: KlaviyoElement[],
    links: {
        self: "string",
        first: "string",
        last: "string",
        prev: "string",
        next: "string"
    }
};
