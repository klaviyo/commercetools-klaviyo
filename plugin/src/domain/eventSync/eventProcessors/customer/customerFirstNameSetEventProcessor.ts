import { AbstractEventProcessor } from '../abstractEventProcessor';
import logger from '../../../../utils/log';
import { CustomerFirstNameSetMessage } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/message';

export class CustomerFirstNameSetEventProcessor extends AbstractEventProcessor {
    isEventValid(): boolean {
        const message = this.ctMessage as unknown as CustomerFirstNameSetMessage;
        return message.resource.typeId === 'customer' && message.type === 'CustomerFirstNameSet';
    }

    generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as CustomerFirstNameSetMessage;
        logger.info(`processing CT ${message.type} message`);
        const body: ProfileRequest = {
            data: {
                type: 'profile',
                attributes: {
                    first_name: message.firstName,
                },
                meta: {
                    identifiers: {
                        external_id: message.resource.id,
                    },
                },
            },
        };
        return Promise.resolve([
            {
                body,
                type: 'profileUpdated',
            },
        ]);
    }
}
