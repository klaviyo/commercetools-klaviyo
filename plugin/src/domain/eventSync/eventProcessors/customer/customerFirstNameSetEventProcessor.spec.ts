import { MessageDeliveryPayload } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/subscription';
import { mock, mockDeep } from 'jest-mock-extended';
import { expect as exp } from 'chai';
import { CustomerFirstNameSetEventProcessor } from './customerFirstNameSetEventProcessor';
import { getSampleCustomerFirstNameSetMessage } from '../../../../test/testData/ctCustomerMessages';
import { Context } from "../../../../types/klaviyo-context";

const context: Context = mock<Context>();

describe('CustomerFirstNameSetEventProcessor > isEventValid', () => {
    it('should return valid when the customer name set event has all the required fields', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'customer' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'CustomerFirstNameSet' });

        const event = CustomerFirstNameSetEventProcessor.instance(ctMessageMock, context);

        exp(event.isEventValid()).to.be.true;
    });

    it.each`
        resource      | type
        ${'invalid'}  | ${'CustomerFirstNameSet'}
        ${'customer'} | ${'invalid'}
    `('should return invalid when is not a valid customer first name set message', ({ resource, type }) => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: resource } });
        Object.defineProperty(ctMessageMock, 'type', { value: type });

        const event = CustomerFirstNameSetEventProcessor.instance(ctMessageMock, context);

        exp(event.isEventValid()).to.be.false;
    });
});

describe('CustomerFirstNameSetEventProcessor > generateKlaviyoEvent', () => {
    it('should generate the klaviyo event when the input customer first name set event is valid', async () => {
        const inputMessage = getSampleCustomerFirstNameSetMessage();
        const message = inputMessage as unknown as MessageDeliveryPayload;
        const event = CustomerFirstNameSetEventProcessor.instance(message, context);

        const klaviyoEvent = await event.generateKlaviyoEvents();
        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });
});
