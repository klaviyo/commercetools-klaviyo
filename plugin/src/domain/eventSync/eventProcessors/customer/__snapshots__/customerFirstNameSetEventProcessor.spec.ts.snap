// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerFirstNameSetEventProcessor > generateKlaviyoEvent should generate the klaviyo event when the input customer first name set event is valid 1`] = `
{
  "data": {
    "attributes": {
      "first_name": "Bob",
    },
    "meta": {
      "identifiers": {
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
      },
    },
    "type": "profile",
  },
}
`;
